rfcs/rfc-0000-discriminators.md

```diff
@@ -0,0 +1,123 @@
+it stands to solve multiple problems when implemented correctly:
+
+- Users should in most, if not all, cases be able to pick their chosen username.
+- This should curb any incentive to sell / steal accounts for "rare usernames".
+
+  Certain "rare" discriminators can also be prohibited.
+
+- Allow users who do not want to be easily discoverable to stay hidden.
+
+  Just a 4-digit discriminator provides 10,000 unique users.
+  Expanding on this, we could for example, let users hide discriminators in servers and similar privacy features.
+
+# Guide-level explanation
+
+Revolt would be switching to a new username system made up of three parts:
+
+```
+Hello, World!#123456
+^ your chosen name
+             ^ separator between chosen name and discriminator
+              ^ discriminator
+```
+
+Discriminators for new users are generated randomly on sign up while existing users will
+get a randomly generated discriminators upon the server software being upgraded.
+
+> Revolt treats usernames as case-insensitive so a full username such as "Jason#1234" is
+> effectively the same as "jasON#1234" but only "Jason#1234" will display in the UI.
+
+Users may:
+
+- Change the case of their username at any moment, such as from "Jason" to "jasON".
+
+  This will never affect their discriminator.
+
+- Change their username, but may have their discriminator regenerated if there is a conflict.
+
+  Such as if you are changing from "Jason#1234" to "Phil", but "Phil#1234" is taken.
+
+### User Opinion
+
+Prior discussions on the forums primarily seem to lean in favour of discriminators being
+implemented into Revolt, the internal team has not raised any complaints against it, and
+the primary community on Revolt ("Revolt Lounge") appears to be somewhat split on the
+issue.
+
+As a recent case study, Discord's change to unique usernames only has generated a lot of
+backlash and criticism from the community, it appears to be overwhemingly negative but
+these experiences are mostly coming from long-term users of Discord. The true opinion of
+the average user has not been gauged yet.
+
+### Usability
+
+This may make it more difficult to understand how to add users on Revolt, however this
+shouldn't be an issue as long as it's clearly communicated through the user interface.
+Revolt already allows a wide range of Unicode in usernames so we are definitely not
+making a significant UX hit towards the usability of the friend system.
+
+> There are further considerations listed at the end of this document which must be
+> addressed if this RFC were to be implemented.
+
+# Reference-level explanation
+
+WIP
+
+This is the technical section of the RFC, it should go over in detail:
+
+- Its interaction with other features
+- How this will be implemented
+- Corner or edge cases
+
+This section should reference the examples in the previous section and disect them in more detail.
+
+# Drawbacks
+
+WIP
+
+Why should this not be added.
```

Comment:
Since Discord believes discriminators are no longer viable for their platform, I think it'd probably be worth looking at the reasons they list on their blog post, which are supposed to be solved by implementing a more traditional username system.

https://discord.com/blog/usernames

The biggest issue that stands out to me is that, supposedly, **users are often found to be forgetting their discriminators, or they do not even know what they are.**

I will admit I've almost never been able to remember my own discriminators on Discord, but it has literally *never* mattered to me. Whenever I have to share my Discord username with someone, I can always just... open Discord and *look at it.* Most of the time I can even just copy it directly from there.

Apart from that, I can definitely see at least *some* people getting confused about what these random numbers after your actual username are.... but *surely* it'd be easier to just add some kind of tooltip in the UI explaining it?? Or even just straight up tell the user on sign-up. Like, something along the lines of, "Here's your new username! Be sure to give this entire string of characters when sharing it with your friends!"

Reply:
I've known my discriminator since the last time I changed it 5-6 years ago (mainly because I put it on a lot of the new accounts I make) but I can definitely see someone forgetting about it. But like you said, it's something you can just *look at* and one that you can even customize.

Also, the vast majority of Revolt users come from Discord, so they would understand the concept immediately. If not, then like you said, there can be a tooltip. (Or even something in the docs)

Reply:
I realize a lot of discussions around Revolt development are going to inevitably mention Discord in one way or another, but I honestly believe it'd be in best interest to not automatically assume people are inherently familiar with that platform. It's very well within the realm of possibility that people can come to this platform without having made a Discord account, and perhaps there are even people who have managed to hear about Revolt before Discord.

Reply:
I've written up a discussion around the Discord case study here: https://github.com/insertish/revolt-rfcs/blob/discriminators/rfcs/rfc-0000-discriminators.md#drawbacks

I'm going to resolve this conversation, please re-open on relevant points if there's something more to be said about it.